File: swift/src/Revparse.swift
/// Commit-oriented revision parsing for common Git revision expressions.
/// Parity target: libgit2 `git_revparse_single` / `git_revparse`
import Foundation

public struct RevSpec: Sendable, Equatable {
    public let from: OID?
    public let to: OID?
    public let isRange: Bool
    public let usesMergeBase: Bool

    public init(from: OID?, to: OID?, isRange: Bool, usesMergeBase: Bool) {
        self.from = from
        self.to = to
        self.isRange = isRange
        self.usesMergeBase = usesMergeBase
    }
}

/// Resolve a common revision expression to a commit OID.
///
/// Supported subset:
/// - full OIDs
/// - refs and short refs like `main`, `tags/v1`, `origin/main`
/// - `HEAD^`, `HEAD^N`, `HEAD~N`
public func resolveRevision(gitDir: String, spec: String) throws -> OID {
    let trimmed = spec.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
        throw MuonGitError.invalidSpec("empty revision spec")
    }
    if trimmed.contains("...") || isTwoDotRange(trimmed) {
        throw MuonGitError.invalidSpec("range '\(trimmed)' does not resolve to a single revision")
    }

    let (baseSpec, suffix) = try splitBaseAndSuffix(trimmed)
    var current = try readObject(gitDir: gitDir, oid: resolveRevisionBase(gitDir: gitDir, spec: baseSpec))

    let chars = Array(suffix)
    var index = 0
    while index < chars.count {
        switch chars[index] {
        case "~":
            index += 1
            let start = index
            while index < chars.count, chars[index].wholeNumberValue != nil {
                index += 1
            }
            let count = start == index ? 1 : Int(String(chars[start..<index])) ?? -1
            guard count >= 0 else {
                throw MuonGitError.invalidSpec("invalid ancestry operator in '\(trimmed)'")
            }
            for _ in 0..<count {
                let commit = try peelRevisionCommit(gitDir: gitDir, object: current, spec: trimmed)
                guard let parent = commit.parentIds.first else {
                    throw MuonGitError.invalidSpec("revision '\(trimmed)' has no first parent")
                }
                current = try readObject(gitDir: gitDir, oid: parent)
            }
        case "^":
            index += 1
            let start = index
            while index < chars.count, chars[index].wholeNumberValue != nil {
                index += 1
            }
            let parentIndex = start == index ? 1 : Int(String(chars[start..<index])) ?? -1
            guard parentIndex >= 0 else {
                throw MuonGitError.invalidSpec("invalid parent selector in '\(trimmed)'")
            }
            let commit = try peelRevisionCommit(gitDir: gitDir, object: current, spec: trimmed)
            if parentIndex == 0 {
                current = try readObject(gitDir: gitDir, oid: commit.oid)
                continue
            }
            guard commit.parentIds.count >= parentIndex else {
                throw MuonGitError.invalidSpec("revision '\(trimmed)' has no parent \(parentIndex)")
            }
            current = try readObject(gitDir: gitDir, oid: commit.parentIds[parentIndex - 1])
        default:
            throw MuonGitError.invalidSpec("unsupported revision syntax '\(trimmed)'")
        }
    }

    return try peelRevisionCommit(gitDir: gitDir, object: current, spec: trimmed).oid
}

public func revparseSingle(gitDir: String, spec: String) throws -> GitObject {
    try readObject(gitDir: gitDir, oid: resolveRevision(gitDir: gitDir, spec: spec))
}

public func revparse(gitDir: String, spec: String) throws -> RevSpec {
    let trimmed = spec.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
        throw MuonGitError.invalidSpec("empty revision spec")
    }

    if let (left, right) = splitRange(trimmed, operator: "...") {
        return RevSpec(
            from: try resolveRevision(gitDir: gitDir, spec: left),
            to: try resolveRevision(gitDir: gitDir, spec: right),
            isRange: true,
            usesMergeBase: true
        )
    }

    if let (left, right) = splitTwoDotRange(trimmed) {
        return RevSpec(
            from: try resolveRevision(gitDir: gitDir, spec: left),
            to: try resolveRevision(gitDir: gitDir, spec: right),
            isRange: true,
            usesMergeBase: false
        )
    }

    return RevSpec(
        from: nil,
        to: try resolveRevision(gitDir: gitDir, spec: trimmed),
        isRange: false,
        usesMergeBase: false
    )
}

func revisionReadCommit(gitDir: String, oid: OID) throws -> Commit {
    let object = try readObject(gitDir: gitDir, oid: oid)
    guard object.objectType == .commit else {
        throw MuonGitError.invalidSpec("revision '\(oid.hex)' is not a commit")
    }
    return try object.asCommit()
}

private func resolveRevisionBase(gitDir: String, spec: String) throws -> OID {
    let trimmed = spec.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
        throw MuonGitError.invalidSpec("missing base revision")
    }

    if looksLikeFullOID(trimmed) {
        let oid = OID(hex: trimmed)
        if (try? readObject(gitDir: gitDir, oid: oid)) != nil {
            return oid
        }
    }

    for candidate in revisionReferenceCandidates(trimmed) {
        if let oid = try? resolveReference(gitDir: gitDir, name: candidate) {
            return oid
        }
    }

    throw MuonGitError.notFound("could not resolve revision '\(trimmed)'")
}

private func peelRevisionCommit(gitDir: String, object: GitObject, spec: String) throws -> Commit {
    let peeled = object.objectType == .tag ? try object.peel(gitDir: gitDir) : object
    guard peeled.objectType == .commit else {
        throw MuonGitError.invalidSpec("revision '\(spec)' does not resolve to a commit")
    }
    return try peeled.asCommit()
}

private func splitBaseAndSuffix(_ spec: String) throws -> (String, Substring) {
    let index = spec.firstIndex { $0 == "^" || $0 == "~" } ?? spec.endIndex
    let base = String(spec[..<index])
    guard !base.isEmpty else {
        throw MuonGitError.invalidSpec("missing base revision in '\(spec)'")
    }
    return (base, spec[index...])
}

private func splitRange(_ spec: String, operator: String) -> (String, String)? {
    guard let range = spec.range(of: `operator`) else {
        return nil
    }
    let left = String(spec[..<range.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
    let right = String(spec[range.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
    guard !left.isEmpty, !right.isEmpty else {
        return nil
    }
    return (left, right)
}

private func splitTwoDotRange(_ spec: String) -> (String, String)? {
    guard !spec.contains("...") else { return nil }
    return splitRange(spec, operator: "..")
}

private func isTwoDotRange(_ spec: String) -> Bool {
    splitTwoDotRange(spec) != nil
}

private func looksLikeFullOID(_ spec: String) -> Bool {
    spec.count == OID.sha1HexLength
        && spec.unicodeScalars.allSatisfy { scalar in
            let value = scalar.value
            return (48...57).contains(value) || (65...70).contains(value) || (97...102).contains(value)
        }
}

private func revisionReferenceCandidates(_ spec: String) -> [String] {
    var candidates = [spec]
    if !spec.hasPrefix("refs/") {
        candidates.append("refs/\(spec)")
        candidates.append("refs/heads/\(spec)")
        candidates.append("refs/tags/\(spec)")
        candidates.append("refs/remotes/\(spec)")
    }
    return candidates
}

File: rust/src/revparse.rs
//! Commit-oriented revision parsing for common Git revision expressions.
//! Parity target: libgit2 `git_revparse_single` / `git_revparse`

use std::path::Path;

use crate::commit::Commit;
use crate::error::MuonGitError;
use crate::object::{read_object, GitObject};
use crate::refs::resolve_reference;
use crate::OID;

/// A parsed revision expression.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RevSpec {
    pub from: Option<OID>,
    pub to: Option<OID>,
    pub is_range: bool,
    pub uses_merge_base: bool,
}

/// Resolve a common revision expression to a commit OID.
///
/// Supported subset:
/// - full OIDs
/// - refs and short refs like `main`, `tags/v1`, `origin/main`
/// - `HEAD^`, `HEAD^N`, `HEAD~N`
pub fn resolve_revision(git_dir: &Path, spec: &str) -> Result<OID, MuonGitError> {
    let trimmed = spec.trim();
    if trimmed.is_empty() {
        return Err(MuonGitError::InvalidSpec("empty revision spec".into()));
    }
    if trimmed.contains("...") || is_two_dot_range(trimmed) {
        return Err(MuonGitError::InvalidSpec(format!(
            "range '{}' does not resolve to a single revision",
            trimmed
        )));
    }

    let (base_spec, suffix) = split_base_and_suffix(trimmed)?;
    let mut current = read_object(git_dir, &resolve_base_oid(git_dir, base_spec)?)?;

    let mut idx = 0usize;
    let suffix_bytes = suffix.as_bytes();
    while idx < suffix_bytes.len() {
        match suffix_bytes[idx] {
            b'~' => {
                idx += 1;
                let start = idx;
                while idx < suffix_bytes.len() && suffix_bytes[idx].is_ascii_digit() {
                    idx += 1;
                }
                let count = if start == idx {
                    1usize
                } else {
                    suffix[start..idx].parse::<usize>().map_err(|_| {
                        MuonGitError::InvalidSpec(format!(
                            "invalid ancestry operator in '{}'",
                            trimmed
                        ))
                    })?
                };
                for _ in 0..count {
                    let commit = peel_to_commit(git_dir, &current, trimmed)?;
                    let parent = commit.parent_ids.first().ok_or_else(|| {
                        MuonGitError::InvalidSpec(format!(
                            "revision '{}' has no first parent",
                            trimmed
                        ))
                    })?;
                    current = read_object(git_dir, parent)?;
                }
            }
            b'^' => {
                idx += 1;
                let start = idx;
                while idx < suffix_bytes.len() && suffix_bytes[idx].is_ascii_digit() {
                    idx += 1;
                }
                let parent_index = if start == idx {
                    1usize
                } else {
                    suffix[start..idx].parse::<usize>().map_err(|_| {
                        MuonGitError::InvalidSpec(format!(
                            "invalid parent selector in '{}'",
                            trimmed
                        ))
                    })?
                };
                let commit = peel_to_commit(git_dir, &current, trimmed)?;
                if parent_index == 0 {
                    current = read_object(git_dir, &commit.oid)?;
                    continue;
                }
                let parent = commit.parent_ids.get(parent_index - 1).ok_or_else(|| {
                    MuonGitError::InvalidSpec(format!(
                        "revision '{}' has no parent {}",
                        trimmed, parent_index
                    ))
                })?;
                current = read_object(git_dir, parent)?;
            }
            _ => {
                return Err(MuonGitError::InvalidSpec(format!(
                    "unsupported revision syntax '{}'",
                    trimmed
                )));
            }
        }
    }

    Ok(peel_to_commit(git_dir, &current, trimmed)?.oid)
}

/// Resolve a common revision expression to a commit object.
pub fn revparse_single(git_dir: &Path, spec: &str) -> Result<GitObject, MuonGitError> {
    let oid = resolve_revision(git_dir, spec)?;
    read_object(git_dir, &oid)
}

/// Parse a common revision expression into either a single commit or a range.
pub fn revparse(git_dir: &Path, spec: &str) -> Result<RevSpec, MuonGitError> {
    let trimmed = spec.trim();
    if trimmed.is_empty() {
        return Err(MuonGitError::InvalidSpec("empty revision spec".into()));
    }

    if let Some((from, to)) = split_range(trimmed, "...") {
        return Ok(RevSpec {
            from: Some(resolve_revision(git_dir, from)?),
            to: Some(resolve_revision(git_dir, to)?),
            is_range: true,
            uses_merge_base: true,
        });
    }

    if let Some((from, to)) = split_two_dot_range(trimmed) {
        return Ok(RevSpec {
            from: Some(resolve_revision(git_dir, from)?),
            to: Some(resolve_revision(git_dir, to)?),
            is_range: true,
            uses_merge_base: false,
        });
    }

    Ok(RevSpec {
        from: None,
        to: Some(resolve_revision(git_dir, trimmed)?),
        is_range: false,
        uses_merge_base: false,
    })
}

pub(crate) fn read_commit(git_dir: &Path, oid: &OID) -> Result<Commit, MuonGitError> {
    read_object(git_dir, oid)?.as_commit().map_err(|_| {
        MuonGitError::InvalidSpec(format!("revision '{}' is not a commit", oid.hex()))
    })
}

fn resolve_base_oid(git_dir: &Path, spec: &str) -> Result<OID, MuonGitError> {
    let trimmed = spec.trim();
    if trimmed.is_empty() {
        return Err(MuonGitError::InvalidSpec("missing base revision".into()));
    }

    if looks_like_full_oid(trimmed) {
        if let Ok(oid) = OID::from_hex(trimmed) {
            if read_object(git_dir, &oid).is_ok() {
                return Ok(oid);
            }
        }
    }

    for candidate in reference_candidates(trimmed) {
        if let Ok(oid) = resolve_reference(git_dir, &candidate) {
            return Ok(oid);
        }
    }

    Err(MuonGitError::NotFound(format!(
        "could not resolve revision '{}'",
        trimmed
    )))
}

fn peel_to_commit(
    git_dir: &Path,
    object: &GitObject,
    spec: &str,
) -> Result<Commit, MuonGitError> {
    let peeled = if object.obj_type == crate::ObjectType::Tag {
        object.peel(git_dir)?
    } else {
        object.clone()
    };

    peeled.as_commit().map_err(|_| {
        MuonGitError::InvalidSpec(format!(
            "revision '{}' does not resolve to a commit",
            spec
        ))
    })
}

fn split_base_and_suffix(spec: &str) -> Result<(&str, &str), MuonGitError> {
    let first_suffix = spec.find(['^', '~']).unwrap_or(spec.len());
    let base = &spec[..first_suffix];
    if base.is_empty() {
        return Err(MuonGitError::InvalidSpec(format!(
            "missing base revision in '{}'",
            spec
        )));
    }
    Ok((base, &spec[first_suffix..]))
}

fn split_range<'a>(spec: &'a str, operator: &str) -> Option<(&'a str, &'a str)> {
    let idx = spec.find(operator)?;
    let left = spec[..idx].trim();
    let right = spec[idx + operator.len()..].trim();
    if left.is_empty() || right.is_empty() {
        return None;
    }
    Some((left, right))
}

fn split_two_dot_range(spec: &str) -> Option<(&str, &str)> {
    if spec.contains("...") {
        return None;
    }
    split_range(spec, "..")
}

fn is_two_dot_range(spec: &str) -> bool {
    split_two_dot_range(spec).is_some()
}

fn looks_like_full_oid(spec: &str) -> bool {
    spec.len() == 40 && spec.as_bytes().iter().all(u8::is_ascii_hexdigit)
}

fn reference_candidates(spec: &str) -> Vec<String> {
    let mut candidates = Vec::new();
    candidates.push(spec.to_string());

    if !spec.starts_with("refs/") {
        candidates.push(format!("refs/{}", spec));
        candidates.push(format!("refs/heads/{}", spec));
        candidates.push(format!("refs/tags/{}", spec));
        candidates.push(format!("refs/remotes/{}", spec));
    }

    candidates
}
